tools/rg_generate_shaders_gl.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+"""
+Converts a fragment or vertex shader file into a GL shader file.
+
+You can use it like this:
+
+    $ ./rg_generate_shaders_gl.py --cmake-output \\
+        ../src/rendergraph/shaders/generated_shaders_gl.cmake \\
+        ../src/rendergraph/shaders/*.{vert,frag}
+"""
+import argparse
+import logging
+import os
+import pathlib
+import re
+import shutil
+import subprocess
+import tempfile
+import typing
+
+
+def find_executable(
+    executable_name: str, additional_paths: list[str] | None = None
+) -> pathlib.Path:
+    """Find an executable by name in $PATH and in the additional paths."""
+    if executable_path := shutil.which(executable_name):
+        return pathlib.Path(executable_path)
+
+    if additional_paths:
+        if executable_path := shutil.which(
+            executable_name, path=os.pathsep.join(additional_paths)
+        ):
+            return pathlib.Path(executable_path)
+
+    raise OSError(f"Executable {executable_name!r} not found!")
+
+
+QSB_EXECUTABLE = find_executable(
+    "qsb",
+    additional_paths=[
+        "/usr/lib/qt6/bin",
+        "/lib/qt6/bin",
+        "/usr/local/lib/qt6/bin",
+    ],
+)
+
+
+def parse_shader(input_filepath: pathlib.Path) -> typing.Iterator[str]:
+    """Parse a Fragment/Vertex shader file and yield lines for a GL file."""
+    with tempfile.NamedTemporaryFile() as fp:
+        subprocess.check_call(
+            [
+                QSB_EXECUTABLE,
+                "--glsl",
+                "120",
+                "--output",
+                fp.name,
+                input_filepath,
+            ]
+        )
+        output = subprocess.check_output(
+            [QSB_EXECUTABLE, "--dump", fp.name],
+            encoding="utf-8",
+            universal_newlines=True,
+        )
+
+    comment_added = False
+    ok = False
+    in_shader_block = 0
+    buffered_blank_line = False
+    for line in output.splitlines():
+        if in_shader_block == 2:
+            if re.match(r"^\*\*", line):
+                ok = True
+            else:
+                if not comment_added and not re.match(r"^#", line):
+                    yield "//// GENERATED - EDITS WILL BE OVERWRITTEN"
+                    comment_added = True
+                if line:
+                    if buffered_blank_line:
+                        yield ""
+                        buffered_blank_line = False
+                    yield line
+                else:
+                    buffered_blank_line = True
+        elif in_shader_block == 1:
+            if line.rstrip() == "Contents:":
+                in_shader_block = 2
+        else:
+            if line.rstrip() == "Shader 1: GLSL 120 [Standard]":
+                in_shader_block = 1
```

Comment:
Very naive suggestion here, to help flattening the nested logic

```suggestion
        if in_shader_block == 2:
            if re.match(r"^\*\*", line):
                ok = True
            else:
                if not comment_added and not re.match(r"^#", line):
                    yield "//// GENERATED - EDITS WILL BE OVERWRITTEN"
                    comment_added = True
                if line:
                    if buffered_blank_line:
                        yield ""
                        buffered_blank_line = False
                    yield line
                else:
                    buffered_blank_line = True
        elif in_shader_block == 1 and line.rstrip() == "Contents:":
                in_shader_block = 2
        elif line.rstrip() == "Shader 1: GLSL 120 [Standard]":
                in_shader_block = 1
```